src/python/pants/backend/jvm/rules/targets.py

```diff
@@ -0,0 +1,808 @@
+        JunitConcurrency,
+        JunitNumThreads,
+    )
+
+
+# -----------------------------------------------------------------------------------------------
+# `jvm_app` target
+# -----------------------------------------------------------------------------------------------
+
+# TODO: create common code for App types, including the Bundle field.
+class JvmApp(Target):
+    alias = "jvm_app"
+    core_fields = ()
+
+
+# -----------------------------------------------------------------------------------------------
+# `jvm_binary` target
+# -----------------------------------------------------------------------------------------------
+
+
+class JvmBinaryMain(StringField):
+    """The name of the `main` class, e.g. 'org.pantsbuild.example.hello.main.HelloMain'.
+
+    This class may be present as the source of this target or depended-upon library.
+    """
+
+    alias = "main"
+
+
+class JvmBinaryBasename(StringField):
+    """Base name for the generated `.jar` file, e.g. 'hello'.
+
+    By default, this uses the `name` field. Note this default is unsafe because of the possible
+    conflict when multiple binaries are built.
+    """
+
+    alias = "basename"
+
+
+class JvmBinaryDeployExcludes(JvmExcludes):
+    """A list of `exclude` objects to apply at deploy time.
+
+    If you, for example, deploy a Java servlet that has one version of `servlet.jar` onto a Tomcat
+    environment that provides another version, they might conflict. `deploy_excludes` gives you a
+    way to build your code but exclude the conflicting `jar` when deploying.
+    """
+
+    alias = "deploy_excludes"
+
+
+class JvmBinaryDeployJarRules(PrimitiveField):
+    """A `jar_rules` object for packaging this binary in a deploy JAR."""
+
+    alias = "deploy_jar_rules"
+    value: Optional[JarRules]
+    default = None
+
+    @classmethod
+    def compute_value(
+        cls, raw_value: Optional[JarRules], *, address: Address
+    ) -> Optional[JarRules]:
+        value_or_default = super().compute_value(raw_value, address=address)
+        if value_or_default is not None and not isinstance(value_or_default, JarRules):
+            raise InvalidFieldTypeException(
+                address, cls.alias, value_or_default, expected_type="a `jar_rules` object"
+            )
+        return value_or_default
+
+
+class JvmBinaryManifestEntries(DictStringToStringField):
+    """A dict that specifies entries for `ManifestEntries` for adding to MANIFEST.MF when packaging
+    this binary."""
+
+    alias = "manifset_entries"
+
+
+class JvmBinaryShadingRules(PrimitiveField):
+    """A list of shading rules to apply when building a shaded (aka monolithic aka fat) binary jar.
+
+    The order of the rules matters: the first rule which matches a fully-qualified class name is
+    used to shade it. See shading_relocate(), shading_exclude(), shading_relocate_package(), and
+    shading_exclude_package().
+    """
+
+    alias = "shading_rules"
+    value: Tuple[Any, ...]
```

Comment:
This feels like it probably should have a more specific type in here?

Reply:
It should, but there's no common superclass :/ Some of the objects are `UnaryRule`, some `RelocateRule`, and I think there are even more than that.

We could use `Union` to represent this, but I was tired of writing these bindings so went with `Any`, reasoning this is still an improvement from V1 `targets` because now `target-types2` will at least render that this is a sequence.

Reply:
I'll fix this now that it's a new day and I have the energy to, hehe.